torch/quantization/ns/graph_matcher.py

```diff
@@ -0,0 +1,315 @@
+            nniqat.ConvBn2d,
+        ]),
+        # linear modules
+        set([
+            nn.Linear,
+            nnq.Linear,
+            nnqat.Linear,
+        ]),
+        # linear functionals
+        set([
+            F.linear,
+            toq.linear,
+        ]),
+        # add
+        set([
+            torch.add,
+            toq.add,
+            operator.add,  # x + y
+        ]),
+    ]
+
+    type_a_related_to_b: Set[Tuple[Callable, Callable]] = set()
+
+    for s in sets_of_related_ops:
+        s_list = list(s)
+        # add every bidirectional pair
+        for idx_0 in range(0, len(s_list) - 1):
+            for idx_1 in range(idx_0 + 1, len(s_list)):
+                type_a_related_to_b.add((s_list[idx_0], s_list[idx_1]))
+                type_a_related_to_b.add((s_list[idx_1], s_list[idx_0]))
+
+    return type_a_related_to_b
+
+def get_non_matchable_functions() -> Set[Callable]:
+    """
+    `call_function` nodes pointing to these functions are non-matchable.
+    """
+    # TODO(future PR): allow customizations
+    return set([
+        torch.quantize_per_tensor,
+    ])
+
+def get_non_matchable_modules() -> Set[Callable]:
+    """
+    `call_module` nodes pointing to instances of these types are non-matchable.
+    """
+    # TODO(future PR): allow customizations
+    return set([
+        torch.quantization.ObserverBase,
+        torch.quantization.FakeQuantizeBase,
+    ])
+
+def _getattr_from_fqn(gm: GraphModule, fqn: str) -> Any:
+    """
+    Given a gm and a fqn such as "foo.bar.baz", returns gm.foo.bar.baz.
+    """
+    fqn_parts = fqn.split(".")
+    cur_val = gm
+    for part in fqn_parts:
+        cur_val = getattr(cur_val, part)
+    return cur_val
+
+class _NSGraphMatchableNodesIterator:
+    """
+    Iterates through the graph of gm, starting with the output nodes
+    and continuing backwards.
+    1. Returns matchable nodes, in order
+    2. Skips over non-matchable nodes
+    """
+    def __init__(
+        self,
+        gm: GraphModule,
+        non_matchable_functions: Set[Callable],
+        non_matchable_modules: Set[Callable],
+    ):
+        self.gm: GraphModule = gm
+        self.non_matchable_functions: Set[Callable] = non_matchable_functions
+        self.non_matchable_modules: Set[Callable] = non_matchable_modules
+        self.seen_nodes: Set[Node] = set()
+        self.stack: List[Node] = []
+        for start_node in _get_output_nodes(self.gm.graph):
+            self.stack.append(start_node)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self) -> Node:
+        """
+        Returns the next matchable node.
+        """
+        while len(self.stack) > 0:
+            cur_node = self.stack.pop()
+            if cur_node in self.seen_nodes:
+                continue
+            self.seen_nodes.add(cur_node)
+            # add args of previous nodes to stack
+            # TODO(future PR): handle kwargs as needed
+            for arg in cur_node.args:
```

Comment:
For non matchable nodes, we may need to do special handling when we append inputs to the stack.

Reply:
hmm, would you mind providing more context?  Not sure what you mean.

Let's say a is matchable and b is non_matchable.  If we have a graph of

```
a0 -> b0 -> a1 -> b1
```

And we iterate from b1 backwards, the current code will add a1 to the stack when we are in b1, and add a0 to the stack when we are in b0, etc.